src/rgw/rgw_d3n_datacache.cc

```diff
@@ -235,27 +239,33 @@ void D3nDataCache::put(bufferlist& bl, unsigned int len, std::string& oid)
     _outstanding_write_size = outstanding_write_size;
   }
   ldout(cct, 20) << "D3nDataCache: Before eviction _free_data_cache_size:" << _free_data_cache_size << ", _outstanding_write_size:" << _outstanding_write_size << ", freed_size:" << freed_size << dendl;
-  while (len >= (_free_data_cache_size - _outstanding_write_size + freed_size)) {
-    ldout(cct, 20) << "D3nDataCache: enter eviction, r=" << r << dendl;
+  while (len > (_free_data_cache_size - _outstanding_write_size + freed_size)) {
```

Comment:
is this the operative change (fixing the issue)?

Reply:
yes, this `while (len >=` to `while (len >` change was necessary in order to be able to write to the cache an object that is exactly the size of the cache, ie if `rgw_d3n_l1_datacache_size` is 4 MiB and `s3cmd put` of a 8MiB object is performed the `while (len >=` kept trying to evict space even thou the cache is empty and can accomodate the 4MiB tail part from 8MiB object exactly

and also the change below is required:
https://github.com/ceph/ceph/blob/8ed8e8cd485c01dd9c5eb94f62dd5d1868881d3a/src/rgw/rgw_d3n_datacache.cc#L252-L258

where `sr` is the number of bytes evicted from the cache, if the cache cache is already empty an `0` will be returned to `sr` from the eviction.
before the fix the `while` loop at 
https://github.com/ceph/ceph/blob/8ed8e8cd485c01dd9c5eb94f62dd5d1868881d3a/src/rgw/rgw_d3n_datacache.cc#L242
kept spinning indefinitely as the check for the condition that there is nothing to evict was incorrect